### PR TITLE
ENH: trim adapters from demuxed seqs

### PR DESCRIPTION
**Pull request by thermokarst**
Note to reviewer, please squash and merge, and use the following commit message:

`ENH: trim adapters from demuxed seqs`

I left the PR as two (or more) commits for review purposes only.

Also, the naming for the Adapter params are a little weird (`adapter`, `forward`, and `anywhere`), to correspond to the long-form `cutadapt` names, but I would be open to suggestions for renaming these here in these methods.

---

~While this PR is pretty orthogonal to https://github.com/qiime2/q2-cutadapt/pull/4, I think it might make more sense to get https://github.com/qiime2/q2-cutadapt/pull/4 merged first, then I can revise this PR based on any changes we make in https://github.com/qiime2/q2-cutadapt/pull/4.~

### CODE REVIEW

File: q2_cutadapt/_trim.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        '--error-rate', str(error_rate),
+        '--times', str(times),
+        '--overlap', str(overlap),
+        '-o', str(trimmed_seqs.path / f_read),
+    ]
+
+    if r_read is not None:
+        cmd += ['-p', str(trimmed_seqs.path / r_read)]
+
+    if adapter_f:
```
Comment (gregcaporaso) [inline q2_cutadapt/_trim.py]: Add `is not None` for this and the following if statements.

File: q2_cutadapt/plugin_setup.py [hunk 2]
```diff
@@ -32,6 +45,150 @@
...
+        'match_adapter_wildcards': Bool,
+    },
+    outputs=[
+        ('trimmed_sequences', SampleData[SequencesWithQuality]),
+    ],
+    input_descriptions={
+        'demultiplexed_sequences': 'The single-end sequences to be trimmed.',
+    },
+    parameter_descriptions={
+        'cores': 'Number of CPU cores to use.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: Is there a value to pass to indicate all cores? If so, could you note that here?

File: q2_cutadapt/plugin_setup.py [hunk 3]
```diff
@@ -32,6 +45,150 @@
...
+        'adapter': 'Sequence of an adapter ligated to the 3\' end. The '
+                   'adapter and any subsequent bases are trimmed. If a `$` '
+                   'is appended, the adapter is only found if it is a suffix '
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: What do you think about: 

_a suffix of the read_ -> _at the end of the read_

That seems clearer for our users, but just a suggestion.

File: q2_cutadapt/plugin_setup.py [hunk 4]
```diff
@@ -32,6 +45,150 @@
...
+                   'of the read.',
+        'front': 'Sequence of an adapter ligated to the 5\' end. The adapter '
+                 'and any preceding bases are trimmed. Partial matches at the '
+                 '5\' end are allowed. If a `^` character is prepended, the '
+                 'adapter is only found if it is a prefix of the read.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: If you agree with my previous comment, you should do this too:

_a prefix of the read_ -> _at the beginning of the read_

File: q2_cutadapt/plugin_setup.py [hunk 5]
```diff
@@ -32,6 +45,150 @@
...
+        'anywhere': 'Sequence of an adapter that may be ligated to the 5\' or '
+                    '3\' end. Both types of matches as described under '
+                    '`adapter` and `front` are allowed. If the first base of '
+                    'the read is part of the match, the behavior is as with '
+                    '`front`, otherwise as with `adapter`. This option is '
+                    'mostly for rescuing failed library preparations - do '
+                    'not use if you know which end your adapter was ligated '
+                    'to.',
+        'error_rate': 'Maximum allowed error rate.',
+        'indels': 'Allow indels in alignments.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: _indels in alignments_ -> _insertions of deletions of bases when matching adapters_

File: q2_cutadapt/plugin_setup.py [hunk 6]
```diff
@@ -32,6 +45,150 @@
...
+        'times': 'Remove up to `times` adapters from each read.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: Remove multiple occurrences of an adapter if it is repeated, up to `times` times.

File: q2_cutadapt/plugin_setup.py [hunk 7]
```diff
@@ -32,6 +45,150 @@
...
+        'overlap': 'Require `overlap` overlap between read and adapter for '
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: Require at least `overlap` bases of overlap between...

File: q2_cutadapt/plugin_setup.py [hunk 8]
```diff
@@ -32,6 +45,150 @@
...
+                   'an adapter to be found.',
+        'match_read_wildcards': 'Interpret IUPAC wildcards in reads.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: _IUPAC wildcards_ -> _IUPAC wildcards (e.g., N)_

File: q2_cutadapt/plugin_setup.py [hunk 9]
```diff
@@ -32,6 +45,150 @@
...
+        'match_adapter_wildcards': 'Interpret IUPAC wildcards in adapters.',
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: _IUPAC wildcards_ -> _IUPAC wildcards (e.g., N)_

File: q2_cutadapt/tests/test_trim.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+    # This test is really just to make sure that the command runs - the
+    # detailed tests in the Util Tests below ensure the commands are crafted
+    # appropriately.
+    def test_typical(self):
+        demuxed = Artifact.import_data('SampleData[SequencesWithQuality]',
+                                       self.get_data_path('single-end'))
+        with redirected_stdio(stdout=os.devnull):
+            self.plugin.methods['trim_single'](demuxed)
+        # TODO: test results
```
Comment (gregcaporaso) [inline q2_cutadapt/tests/test_trim.py]: I think you're adding these tests now? They can be pretty minimal - remember, your goal isn't to test the inner-workings of `cutadapt`, just that you're calling it correctly, that it's doing something when you call it, and that the output files are what you expect. For example, you can test that one or more reads are different in a couple of the fastq files - that will ensure that your looping over files is working as expected, and that your command is creating output files that are different from the input files. (When testing the paired-end command, you'll also want to make sure that the output forward reads correspond to the input forward reads, and the output reverse reads correspond to the input reverse reads - in other words, the order didn't get mixed up.)

File: q2_cutadapt/tests/test_trim.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    package = 'q2_cutadapt.tests'
+
+    def setUp(self):
+        super().setUp()
+
+        self.demux_seqs = SingleLanePerSampleSingleEndFastqDirFmt(
+            self.get_data_path('single-end'), mode='r')
+        self.trimmed_seqs = CasavaOneEightSingleLanePerSampleDirFmt()
+
+    def test_build_trim_command_typical(self):
```
Comment (gregcaporaso) [inline q2_cutadapt/tests/test_trim.py]: Does this cover overriding all of the default parameters? I like to do that, to make sure I haven't mixed up and parameter/value pairs, and that all parameters are being passed.

File: q2_cutadapt/plugin_setup.py (hunk 2 above)
Comment (thermokarst) [reply q2_cutadapt/plugin_setup.py]: Not to my knowledge (at least, I didn't see anything in the docs for this).

`# TODO: add missing parallel support libs to the conda recipe;`

File: q2_cutadapt/plugin_setup.py [hunk 12]
```diff
@@ -73,12 +73,13 @@
...
                    'adapter and any subsequent bases are trimmed. If a `$` '
-                   'is appended, the adapter is only found if it is a suffix '
-                   'of the read.',
+                   'is appended, the adapter is only found if it is at the '
+                   'end of the read.',
         'front': 'Sequence of an adapter ligated to the 5\' end. The adapter '
                  'and any preceding bases are trimmed. Partial matches at the '
                  '5\' end are allowed. If a `^` character is prepended, the '
-                 'adapter is only found if it is a prefix of the read.',
+                 'adapter is only found if it is at the beggining of the '
```
Comment (gregcaporaso) [inline q2_cutadapt/plugin_setup.py]: _beggining_ -> _beginning_

**Pull request merged**